### PR TITLE
Introduce a staging environment

### PR DESCRIPTION
**Pull request by LukasWallrich**
This should allow us to call /deploy in any PR to get a test version of the site with that status. Difficult to test without merging, but should not break anything ...

### CODE REVIEW

**Comment by DAKiersz** [comment]
Thanks for proposing that @LukasWallrich, I will take a look.

File: .github/workflows/deploy.yaml
Comment (DAKiersz) [inline .github/workflows/deploy.yaml:141]: Unless I'm mistaken, this will effectively will push changes to the `gh-pages` branch which will trigger deployment against the 'production' host, even with the CNAME file is changed. The domain used for our pages, and the branch that the website is deployed from, is defined in our repository settings (the latter is an outdated method and needs to be updated).

With regards to GitHub pages, you can have one project site for an organisation and unlimited project (repo) sites. However, GitHub pages does not offer having 'multiple environments' for testing in a single project which is very annoying for a collaborate development lifecycle.

I need to take deeper look - but a workaround would be to have a secondary repository (project site host) with another subdomain (test/staging.forrt.org) that will host a staging environment site and domain. This will need some work that I will explore in short-term, as again, we don't have a way to stage in a 'pre-production' environment.

What is interesting is that Netlify hosting had a staging slot to perform exactly what you are proposing here without workarounds, but it is my understanding that we want to stay on GH Pages for now.

I would advice against merging this for now.

Some SOF posts on the matter:
https://github.com/peaceiris/actions-gh-pages?tab=readme-ov-file#%EF%B8%8F-add-cname-file-cname
https://docs.github.com/en/pages/getting-started-with-github-pages/about-github-pages
https://stackoverflow.com/questions/59937478/github-pages-staging-site
https://stackoverflow.com/questions/31911091/creating-staging-and-production-versions-of-a-site-on-github-pages

File: .github/workflows/deploy_pr_on_request.yml
Comment (DAKiersz) [inline .github/workflows/deploy_pr_on_request.yml:13]: An interesting action, but I'm interested to know what advantage does this provide issuing a workflow_dispatch in Actions?

Comment (LukasWallrich) [reply .github/workflows/deploy_pr_on_request.yml:13]: An advantage compared to what? The idea is that this allows the test deploy to be triggered by calling /deploy in a PR. Based on my understanding, this is the easiest way to do it ...

File: .github/workflows/deploy.yaml
Comment (LukasWallrich) [reply .github/workflows/deploy.yaml:141]: The idea is that it deploys to the production environment, to forrt.org/pr-123 - I am not sure if that is fundamentally different to pushing to test.forrt.org, as I don't think that sub-folders can ever interfere with higher level pages on a static webpage? [Agree that the latter would be better, but harder to do.] 

Are you afraid that this would overwrite the production page? I am not 100% sure, but would expect that it would not touch anything outside the destination_dir specified in the action?

File: .github/workflows/deploy_pr_on_request.yml
Comment (DAKiersz) [reply .github/workflows/deploy_pr_on_request.yml:13]: Perhaps I'm missing something here, I can see this will allow a `/deploy` command to be used in PRs.

However, can you not just run your feature branch from https://github.com/forrtproject/forrtproject.github.io/actions/workflows/deploy.yaml by pressing 'Run Workflow' and selecting the branch you want to execute it against? That is what I'm trying to compare against. Assuming the `deploy.yml` is not modified in the PR, the workflow will still checkout the feature branch you selected.

File: .github/workflows/deploy.yaml
Comment (LukasWallrich) [reply .github/workflows/deploy.yaml:141]: Fully agree that this is not best practice but I'd have seen it as a major improvement over the status quo while we stay on GitHu. Nevertheless, it should be easy enough to set up deploy to another repo which would be better -I will give that a go when I can.

File: .github/workflows/deploy_pr_on_request.yml
Comment (LukasWallrich) [reply .github/workflows/deploy_pr_on_request.yml:13]: Sure, that should work - it's just more complicated than comments on the PR. Do you have any concerns about this action?

File: .github/workflows/deploy.yaml
Comment (LukasWallrich) [reply .github/workflows/deploy.yaml:141]: @DAKiersz I now modified the action to push to a staging repository I created - not sure if we need to set up the gh-pages branch there before this works, but the easiest would be to merge and try, given that it no longer touches the production branch - if you are happy with the approach.

File: .github/workflows/deploy_pr_on_request.yml
Comment (DAKiersz) [reply .github/workflows/deploy_pr_on_request.yml:13]: It is usually good practice to avoid 3rd party-actions when necessary, in line with guidance which you can find here https://docs.github.com/en/actions/security-guides/security-hardening-for-github-actions#using-third-party-actions. 

I'm not suggesting this action is malicious, but its just something to bare in mind. Imo remembering commands can be more dauting for inexperienced contributors than clicking a button. Whatever way this is done, I suggest updating the `CONTRIBUTING.md` page to reflect this.

File: .github/workflows/deploy.yaml
Comment (DAKiersz) [reply .github/workflows/deploy.yaml:141]: Speaking from experience - status-quo is magnitude better than interfering with prod. However, as per your updates this should be avoided. I missed that `external_repository` option exists in that action which seems to simplify this process more than expected.

We will need to add the subdomain to our provider and link it to the repository. Myself and @bethaniley work on the this IOI project on Friday's so we can handle to then, and happy to conduct this release for you.

File: .github/workflows/deploy_pr_on_request.yml
Comment (LukasWallrich) [reply .github/workflows/deploy_pr_on_request.yml:13]: /remind me to respond tomorrow [mostly just testing slash command]

**Comment by LukasWallrich** [comment]
/remind me to respond tomorrow [mostly just testing slash command]

File: .github/workflows/deploy_pr_on_request.yml
Comment (LukasWallrich) [reply .github/workflows/deploy_pr_on_request.yml:13]: Here I already tricked myself - if we keep this, it also needs to run on `pull_request_review_comment`

File: .github/workflows/deploy_pr_on_request.yml
Comment (DAKiersz) [inline .github/workflows/deploy_pr_on_request.yml:8]: Suggest we change that to `deploy-pr` which differentiates this from the job in `deploy.yml`.

**Review by DAKiersz — CHANGES_REQUESTED** [review]

File: .github/workflows/deploy_pr_on_request.yml
Comment (LukasWallrich) [reply .github/workflows/deploy_pr_on_request.yml]: Good point - done.

**Comment by DAKiersz** [comment]
I am introducing corrections before testing.

* Staging environment can now be found https://staging.forrt.org/
* Added branch protection for Test to deploy correctly before merging.

File: .github/workflows/deploy_pr_on_request.yml
Comment (DAKiersz) [reply .github/workflows/deploy_pr_on_request.yml:13]: I see we can use a newer version of this workflow. Could you please we raise this ChatOps workflow in a separate PR so we can separate the staging environment from this.

https://github.com/peter-evans/slash-command-dispatch/tree/v4/

Comment (LukasWallrich) [reply .github/workflows/deploy_pr_on_request.yml]: You mean to remove this from this PR? I can't do that today, but feel free to do so - otherwise I will do it asap. 

Thanks for your other work on this!

Comment (LukasWallrich) [reply .github/workflows/deploy_pr_on_request.yml]: Ok, I saw that you already removed it. Thanks! Will do the new PR next week. 

/remind me to raise `/deploy` PR

**Comment by DAKiersz** [comment]
@LukasWallrich A big issue with having a single GitHub staging website with multiple PR builds, is that given that our website artifact is 177 MBs, we will result in reaching the limit of 1GB on our staging environment within few PRs https://docs.github.com/en/pages/getting-started-with-github-pages/about-github-pages. A workflow for automatic clean-ups in the staging repo would potentially remedy this, but again, this will require a bit more effort.

When deployed without `destination_dir`:  https://staging.forrt.org
When deployed on subdirectory: https://staging.forrt.org/134/
Our private repo: https://github.com/forrtproject/webpage-staging/tree/staging/134

I strongly suggest that we get a single staging environment up with concurrency settings under this PR to avoid issues, and look into subdirectories in a separate pull request. Let me know what you think.

**Review by DAKiersz — APPROVED** [review]

**Comment by LukasWallrich** [comment]
Not 100% sure if I understand your suggestions - are you suggesting one staging environment that is overwritten with each PR? We can do that as a first step - and then add a next PR that adds separate subfolders & a cleanup action that deletes all but the last 4. 

However, you could also set this up as is, and I would add the cleanup action latest but Tuesday - until then, this shouldn't become an issue?

**Comment by DAKiersz** [comment]
> Not 100% sure if I understand your suggestions - are you suggesting one staging environment that is overwritten with each PR? We can do that as a first step - and then add a next PR that adds separate subfolders & a cleanup action that deletes all but the last 4.
> 
> However, you could also set this up as is, and I would add the cleanup action latest but Tuesday - until then, this shouldn't become an issue?

For time being yes - website subdirectories did not work as expected and would require more effort due to platform limitations. As per my Slack messages, the first step usually (and certainly for a project this size) is to setup a pre-production/test environment, that we can use to test changes right before merging. This will not stop people from doing local dev work either. Then you try to expand to accommodate ephemeral staging environments. 

I suggested to merge this so we can test your bug #143 in an environment closest to production.

**Comment by DAKiersz** [comment]
We need a second approver for this in order to promote this, but otherwise it LGTM @LukasWallrich :)

**Comment by LukasWallrich** [comment]
@flavioazevedo could you review/approve this so that we can get a staging (testing) environment going?

**Review by flavioazevedo — APPROVED** [review]

**Review by bethaniley — APPROVED** [review]
Can't see any issues in the script and the new staging domain seems to be working 🙂

**Pull request merged**
